[JSC] Deploy more loadPair / storePair in DFG

Constellation opened the pull request:
#### 80e312320f130b264ed3f4694620fd63083f8f38
<pre>
[JSC] Deploy more loadPair / storePair in DFG
<a href="https://bugs.webkit.org/show_bug.cgi?id=246502">https://bugs.webkit.org/show_bug.cgi?id=246502</a>
rdar://101156906

Reviewed by Mark Lam.

This patch deploys more and more loadPair/storePair in DFG.
We also add loadPairPtr and storePairPtr so that we can use them
in pointer storing places too.

* Source/JavaScriptCore/assembler/MacroAssembler.h:
(JSC::MacroAssembler::loadPairPtr):
(JSC::MacroAssembler::storePairPtr):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::emitAllocateRawObject):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::emitStoreStructureWithTypeInfo):
(JSC::AssemblyHelpers::emitAllocateWithNonNullAllocator):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::emitInitializeInlineStorage):
(JSC::AssemblyHelpers::emitInitializeOutOfLineStorage):
(JSC::AssemblyHelpers::emitFillStorageWithJSEmpty):
(JSC::AssemblyHelpers::emitFillStorageWithDoubleEmpty):

Canonical link: <a href="https://commits.webkit.org/255608@main">https://commits.webkit.org/255608@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b966ba958dff7e7a96ef917a1d01cbd51462bcb0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93057 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23657 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102766 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/162991 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2268 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30586 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85440 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98882 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1561 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79531 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28477 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83206 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71589 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84378 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36980 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17108 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/79445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34792 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18322 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/79445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3881 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38664 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40897 "Passed tests") | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/82078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40591 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37516 "Passed tests") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/82078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -356,8 +356,7 @@ void AssemblyHelpers::emitStoreStructureWithTypeInfo(AssemblyHelpers& jit, Trust
 #if USE(JSVALUE64)
-    jit.store32(TrustedImm32(structurePtr->id().bits()), MacroAssembler::Address(dest, JSCell::structureIDOffset()));
-    jit.store32(TrustedImm32(structurePtr->typeInfoBlob()), MacroAssembler::Address(dest, JSCell::indexingTypeAndMiscOffset()));
+    jit.store64(TrustedImm64(static_cast<uint64_t>(structurePtr->id().bits()) | (static_cast<uint64_t>(structurePtr->typeInfoBlob()) << 32)), MacroAssembler::Address(dest, JSCell::structureIDOffset()));
```
This works for little endian, but do we have ports that have big endian uint64_t with the hi and lo 32-bits reversed?

Constellation replied:
It is OK. JIT only supports little endian (there are static_asserts etc.), and we have zero plan to expand it.
We have massive amount of JIT code assuming little endian, and little endian is one of the important pre-condition for JIT. This pre-condition wipes a lot of complicated unnecessary code, so we should not assume non-little endian.

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -626,9 +625,10 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
+
+    ASSERT(static_cast<ptrdiff_t>(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfScrambledHead() + sizeof(void*)) == static_cast<ptrdiff_t>(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfSecret()));
+    loadPairPtr(allocatorGPR, TrustedImm32(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfScrambledHead()), resultGPR, scratchGPR);
+    xorPtr(scratchGPR, resultGPR);
```
This would generate less optimal code for x86_64.  The original xorPtr would just do the xor using a memory location as one of the arguments.  The new case would force it to be loaded into the scratch register first before xor'ing.  Perhaps this `loadPairPtr` approach should only be applies to `CPU(ARM64)`?

Constellation replied:
Added.

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -1962,6 +1962,47 @@ class AssemblyHelpers : public MacroAssembler {
+        move(TrustedImm32(0), scratchGPR);
+#endif
+        for (; pairIndex < pairCount; ++pairIndex)
+            storePair64(emptyValueGPR, emptyValueGPR, baseGPR, TrustedImm32(initialOffset + pairIndex * 2 * sizeof(EncodedJSValue)));
```
While this can work, if the offset ends up being > 128, then this `storePair64` will downgrade to 2 `store64`.  Do we know what the typical `count` is and how often it gets big?  If we can afford another scratch registers to use as the base register or if the baseGPR can be clobbered, the other way to do this is to use the spooler which will adjust the baseGPR as needed.  Alternatively, the underlying ARM64 `stp` instruction supports post increment mode, thought we have not exposed that in the MacroAssembler yet.  If we want to, we can add and use a `storePairPostIncrement64` instead to guarantee that we'll never exceed the 128 offset limit.

Constellation replied:
The count would be smaller. But even if it is wider than that, then still, it is just the same to the old implementation (store64). So this is fine.

MenloDorian commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -1962,6 +1962,47 @@ class AssemblyHelpers : public MacroAssembler {
+        unsigned pairIndex = 0;
+        move(TrustedImm64(bitwise_cast<int64_t>(PNaN)), scratchGPR);
+        for (; pairIndex < pairCount; ++pairIndex)
+            storePair64(scratchGPR, scratchGPR, baseGPR, TrustedImm32(initialOffset + pairIndex * 2 * sizeof(double)));
```
Ditto.

Constellation replied:
Ditto. I don't think we have a large case typically, and even it is large, then it is not a regression (the old code was generating the same thing).

MenloDorian reviewed (approved):
r=me with some comments.

Constellation committed:
[JSC] Deploy more loadPair / storePair in DFG
https://bugs.webkit.org/show_bug.cgi?id=246502
rdar://101156906

Reviewed by Mark Lam.

This patch deploys more and more loadPair/storePair in DFG.
We also add loadPairPtr and storePairPtr so that we can use them
in pointer storing places too.

* Source/JavaScriptCore/assembler/MacroAssembler.h:
(JSC::MacroAssembler::loadPairPtr):
(JSC::MacroAssembler::storePairPtr):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::emitAllocateRawObject):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::emitStoreStructureWithTypeInfo):
(JSC::AssemblyHelpers::emitAllocateWithNonNullAllocator):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::emitInitializeInlineStorage):
(JSC::AssemblyHelpers::emitInitializeOutOfLineStorage):
(JSC::AssemblyHelpers::emitFillStorageWithJSEmpty):
(JSC::AssemblyHelpers::emitFillStorageWithDoubleEmpty):

Canonical link: https://commits.webkit.org/255608@main
